### PR TITLE
ext_proc fuzzer test trigger ENVOY_BUG when clear route cache 

### PR DESCRIPTION
**Pull request by yanjunxiang-google**
This PR is to address an ENVOY_BUG crash exposed by ext_proc fuzzer test reported by:

https://clusterfuzz.corp.google.com/testcase-detail/4756668218736640

The steps to trigger this ENVOY_BUG is below:
1) When handling upstream response body message, ext_proc filter sends the response body to the ext_proc server.
2) The ext_proc server sends back a body response with header mutation and set the clear_route_cache flag.
3) Such clear_route_cache flag will trigger function call into connection_manager to clear route cache, which is set to be blocked in the stage when connection_manager is processing the upstream header response. Thus the ENVOY_BUG is fired.

 

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by KBaichoo** [comment]
/assign @tyxia 

As codeowner

/wait

For CI

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/processor_state.cc:374]:
```suggestion
    // Do not clear the route cache on the response path to downstream client.
```

nit: I think  `response path to downstream client`  is a bit clearer than outbound traffic as it is more widely used and more specific.  WDYT?

**Review by tyxia — CHANGES_REQUESTED** [review]
Thanks for working on this! I think it is the right fix.

Also, cc @wbpcode, I think before PR #26045, clearRouteCache on response path to downstream is just a no-op that won't make any difference rather than triggering the ENVOY_BUG.

File: test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc
Comment (yanjunxiang-google) [inline test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc:2090]: This integration test can reproduce the crash without the change in process_state.cc:clearRouteCache() to bailout for upstream response traffic.

File: source/extensions/filters/http/ext_proc/ext_proc.h
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.h]: Sure, let me remove the counter.   You are right that for outbound traffic there is no need to clear route cache at all.

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/processor_state.cc]: Agreed with the 1st option.

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/processor_state.cc]: Sounds good.

**Comment by yanjunxiang-google** [comment]
This is the fuzzer test case:


ext_proc_data: "scterpc_cre:csp_yilooo0\n\000*!pV1:ae!FoFFF,F\n"

File: test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc
Comment (yanjunxiang-google) [inline test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc:2134]: This is the test code which can trigger the ENVOY_BUG crash.

**Comment by wbpcode** [comment]
> clearRouteCache on response path to downstream client is just a no-op (since route selection is not needed on response path

Nope, the refresh in response phase still may changes the roue and effects the per filter config/metadata... etc that filters may need to use. Apparently it's wrong, so, we should never do this refresh in the response phase.

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (yanavlasov) [inline source/extensions/filters/http/ext_proc/processor_state.cc:370]: I would go with the change in the API as it should not allow clearing route cache when processing response headers. Given that API is still alpha, I think this is a better choice when just coding this restriction in the filter.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/processor_state.cc:370]: Thanks for the comments.  IIUC, you are recommending removing 'clear_route_cache' 
https://github.com/envoyproxy/envoy/blob/0ae464feb4b3d1655baed1087cc97339fc644712/api/envoy/service/ext_proc/v3/external_processor.proto#L291
from 'CommonResponse'
https://github.com/envoyproxy/envoy/blob/0ae464feb4b3d1655baed1087cc97339fc644712/api/envoy/service/ext_proc/v3/external_processor.proto#L245
And put it into a proto message which is ext_proc server response specific to downstream client request.

However, currently the entire 'ProcessingResponse',  sent back by the ext_proc server to Envoy,
https://github.com/envoyproxy/envoy/blob/0ae464feb4b3d1655baed1087cc97339fc644712/api/envoy/service/ext_proc/v3/external_processor.proto#L126

and all its underline fields, including CommonResponse, doesn't differentiate whether it is for the downstream request or upstream response.  I went through the code, and does not find an easy way to make such changes. Please suggest if you have some good ideas.  Alternatively, we can add some comments for this field to mark it as something like "for responding to downstream request only".

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/processor_state.cc:362]: done!

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/processor_state.cc:367]: I am fine with the current way to add a check there.
If you have strong opinion this should be done by have a no-op in the ProcessorState class, and adding an override implementation in the derived class DecodingProcessorState, please let me know. I will change.

File: api/envoy/service/ext_proc/v3/external_processor.proto
Comment (yanavlasov) [inline api/envoy/service/ext_proc/v3/external_processor.proto:290]: Please add a new `RequestHeadersResponse` message and move the `clear_route_cache` flag there. It is better to express the intent through the API definition.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

File: api/envoy/service/ext_proc/v3/external_processor.proto
Comment (yanjunxiang-google) [reply api/envoy/service/ext_proc/v3/external_processor.proto:290]: Thanks for the comments!  I coded it up like below. However, this is a breaking change, and outside of the scope of this fuzzer bug fix.  I am wondering can we do it in a separate PR and give it enough time for comments.

diff --git a/api/envoy/service/ext_proc/v3/external_processor.proto b/api/envoy/service/ext_proc/v3/external_processor.proto
index 5fb9393974..9ec8964048 100644
--- a/api/envoy/service/ext_proc/v3/external_processor.proto
+++ b/api/envoy/service/ext_proc/v3/external_processor.proto
@@ -285,10 +285,19 @@ message CommonResponse {
   // along with the CONTINUE_AND_REPLACE status.
   config.core.v3.HeaderMap trailers = 4;
 
+// This message specifies the response data from the external processing server
+// specially for downstream client request. If the server set this field when
+// responding to the upstream response request, it will be ignored by Envoy.
+  DownstreamRequestResponse downstream_response = 5;
+}
+
+// This message specifies the response data from the external processing server
+// specially for downstream client request.
+message DownstreamRequestResponse {
   // Clear the route cache for the downstream client request. This is necessary
   // if the remote server modified headers that are used to calculate the route.
   // This field is ignored for the upstream response message.
-  bool clear_route_cache = 5;
+  bool clear_route_cache = 1;
 }

Comment (yanjunxiang-google) [reply api/envoy/service/ext_proc/v3/external_processor.proto:290]: filed this issue to track this API change request: https://github.com/envoyproxy/envoy/issues/27772

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/processor_state.cc:433]: nit: decoder_callbacks_ ->downstreamCallbacks()

File: source/extensions/filters/http/ext_proc/processor_state.h
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/processor_state.h:213]: nit: We probably can just have a default {} virtual function here instead of pure virtual.
Then we don' t need have it at all in `EncodingProcessorState`.

I don't feel strongly. It is up to you.

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/processor_state.cc:429]: Could you add back this `ENVOY_LOG(debug, "NOT clearing route cache, it is disabled in the config");`? 

This is meaningful here as it represents the case that ext_proc override the clear setting in common_response and force the clear off.

**Review by tyxia — COMMENTED** [review]
LGTM, modulo nit.

Thanks!

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/processor_state.cc]: done!

File: source/extensions/filters/http/ext_proc/processor_state.h
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/processor_state.h]: sure, done!

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/processor_state.cc:429]: Once it is configured, then this is in normal path, right? So, we will have such log for each response have this flag set.

Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/processor_state.cc:429]: Let me add it back any way.

Comment (tyxia) [reply source/extensions/filters/http/ext_proc/processor_state.cc:429]: The key we originally have log here is because of the override behavior that ext_proc/envoy overrides the ext_proc server's clear setting. This helps with observability/debuggability so that customer can figure out why their clear setting doesn't take effect. Same reason why we have stat here

**Review by tyxia — DISMISSED** [review]
Thanks!

File: api/envoy/service/ext_proc/v3/external_processor.proto
Comment (yanavlasov) [reply api/envoy/service/ext_proc/v3/external_processor.proto:290]: Why can't the #27772 be fixed? What is the point of this intermediate change, that would need to be changed again at some point in the future?

Also it is ok to make breaking changes, since the API is still alpha.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait-any

File: api/envoy/service/ext_proc/v3/external_processor.proto
Comment (yanjunxiang-google) [reply api/envoy/service/ext_proc/v3/external_processor.proto:290]: This breaking API change may trigger lengthy review process about documents and wording, and drag this bug fix for a little bit long.  But if you think we should do it within this PR any way, we can do that.

@stevenzzzz

Comment (yanavlasov) [reply api/envoy/service/ext_proc/v3/external_processor.proto]: Fixing the API the right way will not be lengthy and is not larger than this PR. This speed up PR seem superfluous.

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (stevenzzzz) [inline source/extensions/filters/http/ext_proc/processor_state.cc:438]: do we want to inc this on the encoding path?

Comment (tyxia) [reply source/extensions/filters/http/ext_proc/processor_state.cc:438]: yea, there should be no clearRouteCache in encode/response path. And this is actually what this PR is fixing.

**Comment by yanjunxiang-google** [comment]
Kind ping!

**Comment by stevenzzzz** [comment]
LGTM

**Review by tyxia — APPROVED** [review]

File: api/envoy/service/ext_proc/v3/external_processor.proto
Comment (yanjunxiang-google) [reply api/envoy/service/ext_proc/v3/external_processor.proto]: The  proposed API change is in : https://github.com/envoyproxy/envoy/pull/27852. Even if we can push this API change through,  the logic implemented in this PR probably is still relevant, right? Like only do clear_route_cache() in decoder path.

Comment (yanjunxiang-google) [reply api/envoy/service/ext_proc/v3/external_processor.proto]: BTW,based on a quick look, that API change in https://github.com/envoyproxy/envoy/pull/27852 will trigger significantly code shuffling in onReceiveMessage() since encoding and decoding won't be able to share the same code, like handleHeadersResponse() or handleBodyResponse() any more.

Comment (yanjunxiang-google) [reply api/envoy/service/ext_proc/v3/external_processor.proto]: IMHO, we should get this PR committed to let the ext_proc fuzzer continue run and detecting new issues, if any.

**Comment by yanjunxiang-google** [comment]
Kind ping!

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**
